xds/src/main/java/com/linecorp/centraldogma/xds/k8s/v1/XdsKubernetesService.java

```diff
@@ -78,81 +82,103 @@ public XdsKubernetesService(XdsResourceManager xdsResourceManager) {
+                responseObserver, group, kubernetesEndpointName,
+                K8S_ENDPOINT_AGGREGATORS_DIRECTORY + aggregatorId + ".json",
+                "Create kubernetes endpoint: " + kubernetesEndpointName, aggregator, author, true));
     }
 
-    private void validateWatcherAndPush(
-            StreamObserver<ServiceEndpointWatcher> responseObserver,
-            ServiceEndpointWatcher watcher, Runnable onSuccess) {
+    private void validateKubernetesEndpointAndPush(
+            StreamObserver<KubernetesEndpointAggregator> responseObserver,
+            KubernetesEndpointAggregator aggregator, Runnable onSuccess) {
         // Create a KubernetesEndpointGroup to check if the watcher is valid.
         // We use KubernetesEndpointGroup for simplicity, but we will implement a custom implementation
         // for better debugging and error handling in the future.
         final ContextAwareBlockingTaskExecutor taskExecutor =
                 ServiceRequestContext.current().blockingTaskExecutor();
-        final CompletableFuture<KubernetesEndpointGroup> future =
-                createKubernetesEndpointGroup(watcher, xdsResourceManager.xdsProject().metaRepo(),
-                                              taskExecutor);
-        future.handle((kubernetesEndpointGroup, cause) -> {
-            if (cause != null) {
-                cause = Exceptions.peel(cause);
-                if (cause instanceof IllegalArgumentException || cause instanceof EntryNotFoundException) {
-                    responseObserver.onError(Status.INVALID_ARGUMENT.withCause(cause).asRuntimeException());
-                } else {
-                    responseObserver.onError(Status.INTERNAL.withCause(cause).asRuntimeException());
-                }
-                return null;
-            }
-            final AtomicBoolean completed = new AtomicBoolean();
-            final CompletableFuture<List<Endpoint>> whenReady = kubernetesEndpointGroup.whenReady();
 
-            // Use a schedule to time out the watcher creation until we implement a custom implementation.
-            final ScheduledFuture<?> scheduledFuture = taskExecutor.schedule(() -> {
-                if (!completed.compareAndSet(false, true)) {
-                    return;
-                }
-                kubernetesEndpointGroup.closeAsync();
-                responseObserver.onError(
-                        Status.INTERNAL.withDescription(
-                                "Failed to retrieve k8s endpoints within 5 seconds. watcherName: " +
-                                watcher.getName()).asRuntimeException());
-            }, 5, TimeUnit.SECONDS);
-
-            whenReady.handle((endpoints, cause1) -> {
-                if (!completed.compareAndSet(false, true)) {
+        final ArrayList<CompletableFuture<Void>> futures = new ArrayList<>();
+        for (KubernetesLocalityLbEndpoints kubernetesLocalityLbEndpoints
+                : aggregator.getLocalityLbEndpointsList()) {
```

Comment:
What do you think of returning`INVALID_ARGUMENT` if `aggregator.getLocalityLbEndpointsList().size() == 0`?

Reply:
That's a good suggestion. 👍 👍